core/src/main/java/org/elasticsearch/indices/recovery/RecoverySource.java

```diff
@@ -93,6 +93,14 @@ private RecoveryResponse recover(final StartRecoveryRequest request) throws IOEx
             logger.debug("delaying recovery of {} as source node {} is unknown", request.shardId(), request.targetNode());
             throw new DelayRecoveryException("source node does not have the node [" + request.targetNode() + "] in its state yet..");
         }
+
+        ShardRouting routingEntry = shard.routingEntry();
+        if (request.recoveryType() == RecoveryState.Type.PRIMARY_RELOCATION &&
+            (routingEntry == null || routingEntry.relocating() == false || routingEntry.relocatingNodeId().equals(request.targetNode().getId()) == false)) {
```

Comment:
Looking at the code some more, it can indeed never be null if we get IndexShard through indexService.getShard() as the routing entry is updated before the IndexShard is added to the map of shards in IndexService -> Removed the null check